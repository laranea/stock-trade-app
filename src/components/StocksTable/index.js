// @flow
import React from 'react';

import type { Stock } from '../../types/stockTypes';
import Widget from '../Widget';
import Table from '../Table';

type StocksTableUi = {
  dividendYield: number,
  geometricMean: number,
  peRatio: number,
  stockSymbol: string,
  volumeWeightedStockPrice: number
};

type StocksTableProps = {
  stocksData: Array<Stock>,
  handleGetStocks: () => void
};

const colDefs = [
  'Stock',
  'Dividend Yield',
  'P/E Ratio',
  'Geometric Mean',
  'Volume Weighted Stock Price'
];

export const formatData = (data: Array<Stock>): Array<StocksTableUi> => {
  return data.map(item => ({
    stockSymbol: item.stockSymbol,
    dividendYield: item.dividendYield,
    peRatio: item.peRatio,
    geometricMean: item.geometricMean,
    volumeWeightedStockPrice: item.volumeWeightedStockPrice
  }));
};

export default class StocksTable extends React.Component<StocksTableProps> {
  componentDidMount(): void {
    this.props.handleGetStocks();
  }

  render() {
    return (
      <Widget title="My Stocks">
        <Table colDefs={colDefs} rowData={formatData(this.props.stocksData)} />
      </Widget>
    );
  }
}
