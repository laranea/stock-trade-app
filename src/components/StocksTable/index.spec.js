import React from 'react';
import { shallow } from 'enzyme';

import StocksTable, { formatData } from './';
import { v4 as uuid } from 'uuid';

const MOCK_DATA = [
  {
    dividendYield: null,
    fixedDividend: null,
    geometricMean: null,
    id: uuid(),
    lastDividend: 0,
    parValue: 100,
    peRatio: null,
    stockSymbol: 'TEA',
    type: 'common',
    volumeWeightedStockPrice: null
  },
  {
    dividendYield: null,
    fixedDividend: null,
    geometricMean: null,
    id: uuid(),
    lastDividend: 8,
    parValue: 100,
    peRatio: null,
    stockSymbol: 'POP',
    type: 'common',
    volumeWeightedStockPrice: null
  }
];

const MOCK_FORMATTED_DATA = [
  {
    dividendYield: null,
    geometricMean: null,
    peRatio: null,
    stockSymbol: 'TEA',
    volumeWeightedStockPrice: null
  },
  {
    dividendYield: null,
    geometricMean: null,
    peRatio: null,
    stockSymbol: 'POP',
    volumeWeightedStockPrice: null
  }
];

const handleGetStocksMock = jest.fn();

describe('StocksTable component', () => {
  it('should match snapshot', () => {
    const component = shallow(
      <StocksTable
        stocksData={MOCK_DATA}
        handleGetStocks={handleGetStocksMock}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should call handleGetStocks on load', () => {
    shallow(
      <StocksTable
        stocksData={MOCK_DATA}
        handleGetStocks={handleGetStocksMock}
      />
    );
    expect(handleGetStocksMock).toHaveBeenCalled();
  });

  describe('When calling formattedData', () => {
    it('should return formatted data', () => {
      const result = formatData(MOCK_DATA);
      expect(result).toEqual(MOCK_FORMATTED_DATA);
    });
  });
});
