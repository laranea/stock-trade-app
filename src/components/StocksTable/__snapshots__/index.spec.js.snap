// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StocksTable component should match snapshot 1`] = `
<Widget
  title="My Stocks"
>
  <TableComponent
    colDefs={
      Array [
        "Stock",
        "Dividend Yield",
        "P/E Ratio",
        "Geometric Mean",
        "Volume Weighted Stock Price",
      ]
    }
    rowData={
      Array [
        Object {
          "dividendYield": null,
          "geometricMean": null,
          "peRatio": null,
          "stockSymbol": "TEA",
          "volumeWeightedStockPrice": null,
        },
        Object {
          "dividendYield": null,
          "geometricMean": null,
          "peRatio": null,
          "stockSymbol": "POP",
          "volumeWeightedStockPrice": null,
        },
      ]
    }
  />
</Widget>
`;
